nixos/modules/services/matrix/mautrix-telegram.nix

```diff
@@ -157,35 +215,46 @@ in {
       preStart = ''
         # generate the appservice's registration file if absent
         if [ ! -f '${registrationFile}' ]; then
-          ${pkgs.mautrix-telegram}/bin/mautrix-telegram \
+          ${pkg}/bin/mautrix-telegram \
             --generate-registration \
             --config='${settingsFile}' \
             --registration='${registrationFile}'
         fi
-      '' + lib.optionalString (pkgs.mautrix-telegram ? alembic) ''
-        # run automatic database init and migration scripts
-        ${pkgs.mautrix-telegram.alembic}/bin/alembic -x config='${settingsFile}' upgrade head
       '';
 
       serviceConfig = {
         Type = "simple";
-        Restart = "always";
 
-        ProtectSystem = "strict";
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        NoNewPrivileges = true;
+        PrivateDevices = true;
+        PrivateTmp = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
         ProtectHome = true;
-        ProtectKernelTunables = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
         ProtectKernelModules = true;
-        ProtectControlGroups = true;
+        ProtectKernelTunables = true;
+        ProtectSystem = "strict";
+        Restart = "on-failure";
+        RestartSec = "30s";
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        SystemCallArchitectures = "native";
+        SystemCallErrorNumber = "EPERM";
+        SystemCallFilter = ["@system-service"];
+        UMask = 0027;
 
         DynamicUser = true;
-        PrivateTmp = true;
-        WorkingDirectory = pkgs.mautrix-telegram; # necessary for the database migration scripts to be found
+        WorkingDirectory = pkg; # necessary for the database migration scripts to be found
+        ReadWritePaths = dataDir;
         StateDirectory = baseNameOf dataDir;
```

Comment:
Is it really a good idea to specify the full path of `dataDir` and not just the directory name in `/var/lib`? When `dataDir` is set to a path outside of `/var/lib/`, say for example `/path/to/wa`, then this `StateDirectory` setting will still create `/var/lib/wa` (which will be empty and unused) while `/path/to/wa` will not automatically be created. This will probably cause some problems unless worked around manually with extra effort.

Reply:
Hm. That's right. Yes, I suppose dataDir could be made to always be under /var/lib as you suggested.

Reply:
So I've changed the definition of `dataDir` to be path under /var/lib. From systemd manual it looks like it can be nested as well, just have to be inside of `/var/lib`. What do you think about this?